### ISSUE TITLE
NullPointerException when setting state to ACTIVE

### ISSUE THREAD

**Issue by phillebaba** [graphium-project/graphium-neo4j#3]
I am trying to follow the configurations instructions but am getting stuck when trying to set the state of the imported map. The following request results in a null pointer exception.

```
curl -X PUT "http://localhost:7474/graphium/metadata/graphs/osm/versions/1/state/ACTIVE"
```

```
08:04:24,250 ERROR [GlobalExceptionController.handleInconsistentStateException()]  Error occured during request
java.lang.NullPointerException
	at at.srfg.graphium.neo4j.service.impl.STRTreeCacheManager.update(STRTreeCacheManager.java:199)
	at java.util.Observable.notifyObservers(Observable.java:159)
	at at.srfg.graphium.core.service.impl.GraphVersionMetadataServiceImpl.setGraphVersionState(GraphVersionMetadataServiceImpl.java:116)
	at at.srfg.graphium.api.service.impl.GraphVersionMetadataAPIServiceImpl.setGraphVersionState(GraphVersionMetadataAPIServiceImpl.java:156)
	at at.srfg.graphium.api.service.impl.GraphVersionMetadataAPIServiceImpl.changeState(GraphVersionMetadataAPIServiceImpl.java:216)
	at at.srfg.graphium.api.controller.GraphVersionMetadataApiController.setGraphMetadataFields(GraphVersionMetadataApiController.java:194)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:114)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:963)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:897)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:710)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
```

I am running neo4j in a docker container, and mounting the build plugins to the correct directory.
```
docker run -d -p 7474:7474 -p 7687:7687 -v ${PWD}/plugins:/var/lib/neo4j/plugins -v ${PWD}/conf:/var/lib/neo4j/conf --name neo4j neo4j:3.2.3
```

The issue seems to be with the metadata being passed the observable, but I can't say for sure.

**Comment by phillebaba** [comment]
I solved the issue after some digging through the stack trace. It seems like the `activeIndexCache` is not initialized if the plugin is loaded in a DB that does not contain a single graph.

https://github.com/graphium-project/graphium-neo4j/blob/893e24b9bc97e7827bf42b7e6a56f89d3d8d5543/neo4j-core/src/main/java/at/srfg/graphium/neo4j/service/impl/STRTreeCacheManager.java#L75-L98

This in turn will cause a `NullPointerException` when attempting to activate the graph as the `activeIndexCache` will be called on.

https://github.com/graphium-project/graphium-neo4j/blob/893e24b9bc97e7827bf42b7e6a56f89d3d8d5543/neo4j-core/src/main/java/at/srfg/graphium/neo4j/service/impl/STRTreeCacheManager.java#L199 

The solution for now was to mount a directory to `/var/lib/neo4j/data` so that the neo4j container starts with data before the plugin loads.

Is there any specific reason for not initializing `activeIndexCache` if the DB is empty?

**Comment by micwimmer** [comment]
Do work on the master branch? I think we solved the problem in the develop branch. Could you please check?

**Comment by phillebaba** [comment]
Yes I built everything from the master branch. The PostContruct in STRTreeCacheManager.java will not create a activeIndexCache instance if the graphNamesToIndex is null or empty.

**Comment by micwimmer** [comment]
Please switch to the develop branch, we made some bug fixes there. Currently we are working on a release...

**Comment by micwimmer** [comment]
Last week we released Graphium-Neo4j in v1.0.1. Does it solve your problem?
